cli/tests/integration/npm_tests.rs

```diff
@@ -1546,3 +1548,124 @@ itest!(node_modules_import_check {
+    "Download http://localhost:4545/npm/registry/@denotest/cjs-default-export\n",
+    "Download http://localhost:4545/npm/registry/@denotest/cjs-default-export/1.0.0.tgz\n",
+    "Download http://localhost:4545/npm/registry/@denotest/esm-import-cjs-default/1.0.0.tgz\n",
+  ));
+
+  // modify the package information in the cache to remove the latest version
+  let registry_json_path = "npm/localhost_4545/npm/registry/@denotest/cjs-default-export/registry.json";
+  let mut registry_json: Value =
+    serde_json::from_str(&deno_dir.read_to_string(registry_json_path)).unwrap();
+  remove_version(&mut registry_json, "1.0.0");
+  deno_dir.write(
+    registry_json_path,
+    serde_json::to_string(&registry_json).unwrap(),
+  );
+
+  // should error when `--cache-only` is used now because the version is not in the cache
+  let output = test_context
+    .new_command()
+    .args("run --cached-only main.ts")
+    .run();
+  output.assert_exit_code(1);
+  output.assert_matches_text("error: Could not find npm package '@denotest/cjs-default-export' matching '^1.0.0'.\n");
+
+  // now try running without it, it should download the package now
+  let output = test_context.new_command().args("run main.ts").run();
+  output.assert_matches_text(concat!(
+    "Download http://localhost:4545/npm/registry/@denotest/esm-import-cjs-default\n",
+    "Download http://localhost:4545/npm/registry/@denotest/cjs-default-export\n",
+    "Node esm importing node cjs\n[WILDCARD]",
+  ));
+  output.assert_exit_code(0);
+
+  // now remove the information for the top level package
+  let registry_json_path = "npm/localhost_4545/npm/registry/@denotest/esm-import-cjs-default/registry.json";
+  let mut registry_json: Value =
+    serde_json::from_str(&deno_dir.read_to_string(registry_json_path)).unwrap();
+  remove_version(&mut registry_json, "1.0.0");
+  deno_dir.write(
+    registry_json_path,
+    serde_json::to_string(&registry_json).unwrap(),
+  );
+
+  // should error again for --cached-only
+  let output = test_context
+    .new_command()
+    .args("run --cached-only main.ts")
+    .run();
+  output.assert_exit_code(1);
+  output.assert_matches_text(concat!(
+    "error: Could not find npm package '@denotest/esm-import-cjs-default' matching '1.0.0'. Try retrieving the latest npm package information by running with --reload\n",
+    "    at file:///[WILDCARD]/main.ts:1:8\n",
+  ));
+
+  // now try running without it, it currently will error, but this should work in the future
+  // todo(https://github.com/denoland/deno/issues/16901): fix this
+  let output = test_context.new_command().args("run main.ts").run();
+  output.assert_exit_code(1);
+  output.assert_matches_text(concat!(
+    "error: Could not find npm package '@denotest/esm-import-cjs-default' matching '1.0.0'. Try retrieving the latest npm package information by running with --reload\n",
+    "    at file:///[WILDCARD]/main.ts:1:8\n",
+  ));
+}
```

Comment:
Great test, I really like the reviewing them with the new test infrastructure. So much easier to understand without the boilerplate.

Reply:
It's less arduous to write too!